src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/MLKemOpenSsl.OpenSsl.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.Versioning;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Security.Cryptography
+{
+    public sealed partial class MLKemOpenSsl
+    {
+        private SafeEvpPKeyHandle _key;
+
+        [MemberNotNull(nameof(_key))]
+        private partial void Initialize(SafeEvpPKeyHandle upRefHandle) => _key = upRefHandle;
+
+        public partial SafeEvpPKeyHandle DuplicateKeyHandle()
+        {
+            ThrowIfDisposed();
+            return _key.DuplicateHandle();
+        }
+
+        private static partial MLKemAlgorithm AlgorithmFromHandle(SafeEvpPKeyHandle pkeyHandle, out SafeEvpPKeyHandle upRefHandle)
+        {
+            ArgumentNullException.ThrowIfNull(pkeyHandle);
+            upRefHandle = pkeyHandle.DuplicateHandle();
+
+            try
+            {
+                string name = Interop.Crypto.EvpKemGetName(upRefHandle);
+
+                if (name == MLKemAlgorithm.MLKem512.Name)
```

Comment:
I feel like in prototyping I had this as an internal method on the ParameterSetInfo, so an internal static method on MLKemAlgorithm seems to make sense, and will keep this constrained if we need it in multiple places.  `MLKemAlgorithm.FromName`, or, if you want to be defensive, `FromOpenSslName`.

Reply:
I'm okay with keeping it more generic and reacting later to the situation where we have platforms with differing opinions on "what is a name".